src/piece_picker.cpp

```diff
@@ -2117,7 +2123,12 @@ namespace {
 				}
 				else
 				{
-					for (piece_index_t i = m_cursor; i < m_reverse_cursor; ++i)
+					piece_index_t last_piece = m_reverse_cursor;
```

Comment:
it seems this ought to be just using `m_sequential_end` as the end piece, rather than this complicated logic.
Do you know of any purpose to keep `m_cursor` and `m_reverse_cursor` around? Are they used for anything other than this logic you just replaced?

Reply:
Ok I can do it using the logic you outlined in the earlier comment, but as far as keeping the m_cursor and m_reverse_cursor members around, I think I still need to do that, as after searching through the piece picker file I found references to them outside of this sequential if block, hence I believe they are used in the piece picker for more than sequential download logic.

Reply:
Btw up ahead a couple dozen lines in the code there is this:

```
// in time critical mode, only pick high priority pieces
if (!(options & time_critical_mode))
```

I may be misunderstanding it, but isn't this predicate ensuring that it is *not* in time critical mode?

Reply:
Okay regarding m_cursor and m_reverse_cursor, I looked thoroughly in the piece picker code, and aside from this core pick_pieces() method, they are also found in resize(), we_dont_have(), we_have(), we_have_all(), and set_piece_priority(), to update the cursors, seems to be for bookkeeping purposes. So can I actually safely replace them with their sequential start and end counterparts and just do the same bookkeeping for those?

Reply:
yeah. However, here's a challenge, imagine:

full_torrent_pieces = [0, 100)
sequential_range = [5, 10)

Now, `we_dont_have(4);`. Does that affect sequential_start? presumably not. How about:

```
we_have(5); -> sequential_start = 6
we_dont_have(5); -> sequential_start = ??? (should go back to 5)
```

Somehow the original range needs to be remembered (afaict), just in order to know whether changes to have-state affects it or not.

On the other hand, the `we_dont_have()` feature is fringe and not widely used, perhaps it's OK to ignore those for the purposes of sequential downloading too.

Reply:
> I may be misunderstanding it, but isn't this predicate ensuring that it is not in time critical mode?

That's right. The code that's within that block is picking low-priority pieces. in sequential mode, low priority means picking pieces from the end, i.e. `m_reverse_cursor`

Reply:
> > I may be misunderstanding it, but isn't this predicate ensuring that it is not in time critical mode?
> 
> That's right. The code that's within that block is picking low-priority pieces. in sequential mode, low priority means picking pieces from the end, i.e. `m_reverse_cursor`

But from the code within this if statement, both for the forward and reverse directions, pieces are picked, and not low priority, just anything not the top priority. And the predicate for this if statement seems to be ensuring we're *not* in time critical mode, despite the comment in the line right above it.

Reply:
right, I recall now. only top priority pieces are picked when time-critical is enabled.

And that's done first: https://github.com/arvidn/libtorrent/blob/master/src/piece_picker.cpp#L2076

I'm not sure this is important or improves things. Partly because I don't have any tests that set up the scenarios it's supposed to help with.

Reply:
Ok but this code block you reference is not preceded by a check to make sure time-critical mode is enabled first. It's just run all the time when in sequential download mode. Maybe top priority (7) pieces are only for time critical mode and this code block implicitly assumes as such? Should there be a check to make sure we're in time critical mode before running the code block?